### PR TITLE
Reflection Catcher for Full mode

### PR DESCRIPTION
**Pull request by ardraeiss**
### PURPOSE
Added Reflection Catcher calculations using RIF library for Full mode.

### EFFECT OF CHANGE
- added Reflection Catcher for Full mode for final render.

### TECHNICAL STEPS
- added combined filter;
- used combined filter for non-Legacy modes by final render engine using transparent background and both catcher types.

### CODE REVIEW

File: src/rprblender/engine/image_filter.py [hunk 1]
```diff
@@ -24,6 +24,35 @@
...
+GET_COORD_OR_RETURN(coord, GET_BUFFER_SIZE(outputImage));
+vec4 pixel = ReadPixelTyped(inputImage, coord.x, coord.y);
+vec4 pixel_alpha = ReadPixelTyped(alphaImage, coord.x, coord.y);
+__constant vec4 shadow_color = (vec4)(0.0f, 0.0f, 0.0f, 1.0f);
+vec4 pixel_alpha_neg = pixel_alpha;
+pixel_alpha_neg.x = 1.0f - pixel_alpha.x;
+
+pixel.xyz = pixel.xyz * pixel_alpha.x + shadow_color.xyz * pixel_alpha_neg.x;
+WritePixelTyped(outputImage, coord.x, coord.y, pixel);
+"""
```
Comment (bnagirniak) [inline src/rprblender/engine/image_filter.py:53]: Move this constants to ImageFilterTransparentShadowReflectionCatcher.

**Review by bnagirniak — CHANGES_REQUESTED** [review]
One bug found comparing to RPR1: background is black when "transparent background" option is disabled.

File: src/rprblender/engine/image_filter.py (hunk 1 above)
Comment (bsavery) [reply src/rprblender/engine/image_filter.py:53]: @ardraeiss This change was requested 8 days ago.  Have you addressed?

**Comment by ardraeiss** [comment]
Fixed shadow catcher application.

**Review by bnagirniak — CHANGES_REQUESTED** [review]
After last changes without transparent background it works ok, but with transparent background there are kernel compilation errors:
[ERROR@RIF]: Kernel compilation failed at file C:\JN\WS\RadeonProImageProcessor_Build\src\FilterPass.cpp line 311
Exception in thread Thread-13:
Traceback (most recent call last):
  File "D:\blender\blender-2.92.0-81a0fffb2d2c-windows64\2.92\python\lib\threading.py", line 926, in _bootstrap_inner
    self.run()
  File "D:\blender\blender-2.92.0-81a0fffb2d2c-windows64\2.92\python\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "D:\amd-gpuopen\RadeonProRenderBlenderAddon\src\rprblender\engine\render_engine_2.py", line 79, in do_resolve
    layer_name=self.render_layer_name)
  File "D:\amd-gpuopen\RadeonProRenderBlenderAddon\src\rprblender\engine\engine.py", line 130, in update_render_result
    self._set_render_result(result.layers[0].passes, apply_image_filter, tile_pos, tile_size)
  File "D:\amd-gpuopen\RadeonProRenderBlenderAddon\src\rprblender\engine\engine.py", line 99, in _set_render_result
    self.background_filter.run()
  File "D:\amd-gpuopen\RadeonProRenderBlenderAddon\src\rprblender\engine\image_filter.py", line 109, in run
    self.command_queue.execute()
  File "D:\amd-gpuopen\RadeonProRenderBlenderAddon\src\bindings\pyrpr\src\pyrprimagefilters.py", line 362, in execute
    ContextExecuteCommandQueue(self.context, self, ffi.NULL, ffi.NULL, ffi.NULL)
  File "D:\amd-gpuopen\RadeonProRenderBlenderAddon\src\bindings\pyrpr\src\pyrpr.py", line 74, in wrapped
    raise CoreError(status, f.__name__, argv, module_name)
pyrpr.CoreError: RIF call ContextExecuteCommandQueue(<pyrprimagefilters.Context object at 0x000001C02F212648>, <pyrprimagefilters.CommandQueue object at 0x000001C02F212708>, <cdata 'void *' NULL>, <cdata 'void *' NULL>, <cdata 'void *' NULL>) returned error code <-21> with error message: ''

2021-05-19 09:38:09,599 INFO rpr.RenderEngine [31704]:  Scene synchronization time: 0m 1s 171ms
2021-05-19 09:38:09,600 INFO rpr.RenderEngine [31704]:  Render time: 0m 8s 335ms
[MSG  @RIF]: Loaded RIF API version: 1.7.0.0xbbcfd912
2021-05-19 09:38:31,408 INFO rpr.RenderEngine [28628]:  Scene synchronization time: 0m 1s 523ms
2021-05-19 09:38:31,409 INFO rpr.RenderEngine [28628]:  Render time: 0m 14s 202ms
[MSG  @RIF]: Loaded RIF API version: 1.7.0.0xbbcfd912
C:\Users\bodya\AppData\Local\Temp\\OCL5084T266.cl:964:1: error: vector component access exceeds type 'float4' (vector of 4 'float' values)
DEFINED_FUNCTION;
^~~~~~~~~~~~~~~~
<built-in>:12:820: note: expanded from here
#define DEFINED_FUNCTION /**/int2/**/coord;/**/GET_COORD_OR_RETURN(coord,/**/GET_BUFFER_SIZE(outputImage));/**/vec4/**/pixel/**/=/**/ReadPixelTyped(inputImage,/**/coord.x,/**/coord.y);/**/vec4/**/alpha/**/=/**/ReadPixelTyped(alphaImage,/**/coord.x,/**/coord.y);/**//**//**//**//**//**//**//**//**//**/vec4/**/background/**/=/**/ReadPixelTyped(backgroundImage,/**/coord.x,/**/coord.y);/**//**//**//**//**//**//**//**//**//**//**//**//**//**/vec4/**/reflection/**/=/**/ReadPixelTyped(reflectionImage,/**/coord.x,/**/coord.y);/**/alpha.x/**/+=/**/reflection.x;/**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**//**/pixel.xyz/**/=/**/background.xyz/**/*/**/(1.0f/**/-/**/alpha.x)/**/+/**/pixel.xyz/**/*/**/alpha.x;/**//**//**//**//**//**//**//**//**//**//**//**//**//**/pixel.xyz/**/*=/**/alpha.x;/**/pixel.a/**/=/**/alpha.x;/**//**//**//**//**//**//**//**//**//**//**//**//**//**//**/WritePixelTyped(outputImage,/**/coord.x,/**/coord.y,/**/pixel);/**//**//**//**//**//**//**//**//**/

**Review by bnagirniak — APPROVED** [review]
Tested - now it works ok.

**Pull request merged**